File: Backend/src/main/java/lan/dk/podcastserver/manager/worker/downloader/RTMPDownloader.java
package lan.dk.podcastserver.manager.worker.downloader;


import lan.dk.podcastserver.entity.Item;
import lan.dk.podcastserver.entity.Status;
import lan.dk.podcastserver.repository.ItemRepository;
import lan.dk.podcastserver.repository.PodcastRepository;
import lan.dk.podcastserver.service.MimeTypeService;
import lan.dk.podcastserver.service.ProcessService;
import lan.dk.podcastserver.service.properties.ExternalTools;
import lan.dk.podcastserver.service.properties.PodcastServerParameters;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.annotation.Scope;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.stereotype.Component;

import java.io.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static io.vavr.API.Try;
import static java.util.Objects.nonNull;

@Slf4j
@Component("RTMPDownloader")
@Scope("prototype")
public class RTMPDownloader extends AbstractDownloader {

    private final ProcessService processService;
    private final ExternalTools externalTools;

    int pid = 0;
    Process p = null;

    public RTMPDownloader(ItemRepository itemRepository, PodcastRepository podcastRepository, PodcastServerParameters podcastServerParameters, SimpMessagingTemplate template, MimeTypeService mimeTypeService, ProcessService processService, ExternalTools externalTools) {
        super(itemRepository, podcastRepository, podcastServerParameters, template, mimeTypeService);
        this.processService = processService;
        this.externalTools = externalTools;
    }

    @Override
    public Item download() {
        log.debug("Download");

        try {
            target = getTargetFile(item);
            log.debug("Fichier de sortie : {}" , target.toAbsolutePath().toString());

            p  = processService
                    .newProcessBuilder(externalTools.getRtmpdump(), "-r", getItemUrl(item), "-o", target.toAbsolutePath().toString())
                    .directory(new File("/tmp"))
                    .redirectErrorStream(true)
                    .start();

            pid = processService.pidOf(p);

            new RTMPWatcher(this).run();

            p.waitFor();
            pid = 0;
        } catch (IOException | InterruptedException e) {
            log.error("IOException | InterruptedException :", e);
            stopDownload();
        }
        return item;
    }

    @Override
    public void startDownload() {
        if (pid != 0 && nonNull(p)) { //Relancement du process UNIX
            log.debug("Stop previous process");
            p.destroy();
        }
        super.startDownload();
    }

    @Override
    public void pauseDownload() {
        ProcessBuilder stopProcess = processService.newProcessBuilder("kill", "-STOP", "" + pid);
        Try(stopProcess::start)
            .andThen(super::pauseDownload)
            .onFailure(e -> {
                log.error("IOException :", e);
                this.stopDownload();
            });
    }

    @Override
    public void stopDownload() {
        if (nonNull(p)) {
            p.destroy();
        }
        super.stopDownload();
    }

    @Override
    public Integer compatibility(String url) {
        return url.startsWith("rtmp://") ? 1 : Integer.MAX_VALUE;
    }

    @Slf4j
    @RequiredArgsConstructor
    static class RTMPWatcher implements Runnable {

        static final String DOWNLOAD_COMPLETE = "download complete";
        static final Pattern RTMPDUMP_PROGRESSION_PATTERN_EXTRACTOR = Pattern.compile("[^(]*\\(([0-9]*).*%\\)");
        final RTMPDownloader rtmpDownloader;

        private BufferedReader getBufferedReader(InputStream is) {
            return new BufferedReader(new InputStreamReader(is));
        }

        @Override
        public void run() {
            Process p = rtmpDownloader.p;
            Item item = rtmpDownloader.item;
            Integer pid = rtmpDownloader.pid;

            BufferedReader br = getBufferedReader(p.getInputStream());
            String line;
            log.debug("Reading output of RTMPDump");
            try {
                while ((line = br.readLine()) != null) {
                    log.debug(line);
                    Matcher m = RTMPDUMP_PROGRESSION_PATTERN_EXTRACTOR.matcher(line);
                    if (progressionHasChange(item, m)) {
                        item.setProgression(Integer.parseInt(m.group(1)));
                        rtmpDownloader.convertAndSaveBroadcast();
                    } else if (isDownloadComplete(line)) {
                        log.info("End of download");
                        rtmpDownloader.finishDownload();
                        break;
                    }
                    if (pid == 0) {
                        break;
                    }
                }
            } catch (IOException e) {
                log.error("IOException :", e);
            }
            if (Status.FINISH != item.getStatus() && !rtmpDownloader.stopDownloading.get()) {
                log.debug("Unexpected ending, reset of downloader");
                rtmpDownloader.resetDownload();
            }
        }

        private boolean isDownloadComplete(String ligne) {
            return ligne.toLowerCase().contains(DOWNLOAD_COMPLETE);
        }

        private boolean progressionHasChange(Item item, Matcher m) {
            return m.matches() && Integer.parseInt(m.group(1)) > item.getProgression();
        }
    }
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/download/downloaders/rtmp/RTMPDownloader.kt
package com.github.davinkevin.podcastserver.download.downloaders.rtmp

import com.github.davinkevin.podcastserver.download.ItemDownloadManager
import com.github.davinkevin.podcastserver.download.downloaders.Downloader
import com.github.davinkevin.podcastserver.download.downloaders.DownloaderHelper
import com.github.davinkevin.podcastserver.download.downloaders.DownloadingInformation
import com.github.davinkevin.podcastserver.download.downloaders.DownloadingItem
import com.github.davinkevin.podcastserver.service.ProcessService
import com.github.davinkevin.podcastserver.service.properties.ExternalTools
import org.slf4j.LoggerFactory
import java.io.File
import java.util.*
import java.util.concurrent.atomic.AtomicBoolean
import java.util.regex.Pattern

class RTMPDownloader(
    private val state: DownloaderHelper,
    private val processService: ProcessService,
    private val externalTools: ExternalTools,
) : Downloader {

    private val log = LoggerFactory.getLogger(RTMPDownloader::class.java)

    //* To be removed when Downloader won't be anymore a DownloaderFactory *//
    override val downloadingInformation: DownloadingInformation
        get() = state.info
    override fun with(information: DownloadingInformation, itemDownloadManager: ItemDownloadManager): Downloader =
        throw IllegalAccessException()
    override fun compatibility(downloadingInformation: DownloadingInformation) =
        if (downloadingInformation.urls.size == 1 && downloadingInformation.urls.first().toASCIIString().startsWith("rtmp://")) 1
        else Integer.MAX_VALUE
    //* end of section to remove *//

    private var pid = 0L
    private var p: Process? = null
    private val stopDownloading = AtomicBoolean(false)

    override fun download(): DownloadingItem {
        log.debug("Download")

        try {
            state.target = state.computeTargetFile(downloadingInformation)
            log.debug("out file : {}", state.target.toAbsolutePath().toString())

            val processToExecute = processService
                .newProcessBuilder(
                    externalTools.rtmpdump,
                    "-r",
                    downloadingInformation.url.toASCIIString(),
                    "-o",
                    state.target.toAbsolutePath().toString()
                )
                .directory(File("/tmp"))
                .redirectErrorStream(true)

            p = processService.start(processToExecute)
            pid = processService.pidOf(p!!)

            listenLogs(p!!)

            p!!.waitFor()
            pid = 0
        } catch (e: Exception) {
            throw RuntimeException(e)
        }

        return downloadingInformation.item
    }

    fun listenLogs(process: Process) {
        log.debug("Reading output of RTMPDump")
        var endReached = false
        process.inputStream
            .bufferedReader()
            .lines()
            .peek { log.info("log is: $it") }
            .forEach {
                val (isProgression, progression) = isProgressionLine(it)
                when {
                    isProgression -> broadcastProgression(downloadingInformation.item, progression)
                    isDownloadComplete(it) -> { endReached = true; finishDownload() }
                }
            }

        if (!endReached && !stopDownloading.get()) {
            throw RuntimeException("Unexpected ending, failed download")
        }
    }

    override fun startDownload() {
        stopDownloading.set(false)
        state.startDownload(this, this::failDownload)
    }

    override fun stopDownload() {
        stopDownloading.set(true)
        destroyProcess()
        state.stopDownload()
    }

    private fun destroyProcess() {
        if (p != null) p!!.destroy()
    }

    override fun failDownload() {
        stopDownloading.set(true)
        destroyProcess()
        state.failDownload()
    }

    override fun finishDownload() = state.finishDownload()

    private fun broadcastProgression(item: DownloadingItem, progression: Int) {
        if (item.progression != progression)
            state.broadcast(downloadingInformation)
    }

    private fun isProgressionLine(line: String): Pair<Boolean, Int> {
        val m = RTMP_DUMP_PROGRESSION.matcher(line)
        val isMatch = m.matches()
        return if (isMatch) true to m.group(1).toInt()
        else false to -1
    }

    private fun isDownloadComplete(line: String) = line.lowercase(Locale.getDefault()).contains(DOWNLOAD_COMPLETE)

    companion object {
        private const val DOWNLOAD_COMPLETE = "download complete"
        private val RTMP_DUMP_PROGRESSION = Pattern.compile("[^(]*\\(([0-9]*).*%\\)")!!
    }
}
